### DISCUSSION TITLE
PDF Plumber not extracting tables correctly (text is parsed line by line)

### DISCUSSION THREAD

**Discussion by arthurthlee** [jsvine/pdfplumber#621] [Ask for help with specific PDFs]
Hi,
I'm running into an issue with PDFplumber.

I am attempting to parse this document: https://healthalliance.org/Cms/Media?uri=https%3A%2F%2Fhealthalliance.org%2Fmedia%2Fresources%2Fmed-preauth-drugs.pdf

![image](https://user-images.githubusercontent.com/2807789/157569329-284facb7-71fb-4129-adf7-bc88b5739fad.png)

I am attempting to extract out the "products affected" bullet points and the sections and their text such as "Indications -> All FDA-approved Indications."

I'm using horizontal_strategy: text, vertical_strategy: text, min_words_vertical: 2, and keep_blank_chars: 2.

However, the result of this is something like

```
AAT DEFICIENCY

Products Affected: •Prolastin-c

•Aralast Np INJ 1000MG, 500MG,•Zemaira

800MG

•Glassia



PA Criteria: Criteria Details

Indications: All FDA-approved Indications.

Off-Label Uses: N/A

Exclusion: N/A

Criteria

Required: DOC OF HIGH-RISK PHENOTYPE (E.G. PIZZ,PIZ(NULL),

Medical: PI(NULL)(NULL), PLASMA AAT LEVEL BELOW 11 MICROMOL/L

Information: (CORRESPONDING TO 80M EQUAL TO 35% AND LESS TO COMPLY WITH PROTOCG/D THA OLL) FEV1 GREATER THAN OR N 80% OF PREDICTED ABILITY FOR ADMINISTRATION
```

It looks like PDFPlumber is completely disregarding both the structure in the "Products Affected" section, and the lines in the PA Criteria/Criteria Details section, and printing one line at a time.

Is there something I'm doing wrong?

Thanks

**Comment by jsvine** [comment]
Hi @arthurthlee, converting this from an issue to a discussion, since you're seeking help with a specific PDF. In any case, thanks for your interest in this library. Given that your example PDF/page does contain structural elements (in this case `rect` objects delineating the rows and the separation between the left and right columns), it's often going to be easier to *not* use the `text` strategies.

In the case of your PDF, I'd suggest something along the following code:

```python
p = pdf.pages[1]

# Identify the furthest-left and furthest-right edges 
# of the lines (represented as rects) on the page
rect_x0s = [ r["x0"] for r in p.rects ]
rect_x1s = [ r["x1"] for r in p.rects ]

extracted = p.extract_table({
    "explicit_vertical_lines": [ min(rect_x0s), max(rect_x1s)]
})

print(extracted)
```

This outputs the following:

```
[['Products Affected  •  Prolastin-c   \n'
  '•  Aralast Np INJ 1000MG, 500MG,  •  Zemaira   \n'
  '800MG \n'
  '•  Glassia',
  None],
 ['PA Criteria', 'Criteria Details'],
 ['Indications', 'All FDA-approved Indications.'],
 ['Off-Label Uses', 'N/A'],
 ['Exclusion \nCriteria', 'N/A'],
 ['Required \nMedical \nInformation',
  'DOC OF HIGH-RISK PHENOTYPE (E.G. PIZZ,PIZ(NULL), \n'
  'PI(NULL)(NULL), PLASMA AAT LEVEL BELOW 11 MICROMOL/L \n'
  '(CORRESPONDING TO 80MG/DL) FEV1 GREATER THAN OR \n'
  'EQUAL TO 35% AND LESS THAN 80% OF PREDICTED ABILITY \n'
  'TO COMPLY WITH PROTOCOL FOR ADMINISTRATION'],
 ['Age Restrictions', '18 YEARS OLD OR OLDER'],
 ['Prescriber \nRestrictions', 'N/A'],
 ['Coverage \nDuration', '12 MONTHS AT A TIME'],
 ['Other Criteria',
  'COVERAGE IS LIMITED TO A DOSAGE OF 60MG/KG WEEKLY \n'
  'OR 250MG/KG MONTHLY, B VS D DECISIONS']]
```

... which seems pretty close to what you want.

To see what's happening structurally, you can run the following:

```python
im = p.to_image(resolution=150)

debugged = im.debug_tablefinder({
    "explicit_vertical_lines": [ min(rect_x0s), max(rect_x1s)]
})
```

The representation of `debugged` will be:

![image](https://user-images.githubusercontent.com/534702/157770286-54182b4b-6b64-48db-8346-764551704cb4.png)

**Reply by arthurthlee** [reply]
Thank you, @jsvine  this helps a lot!
As an extension to this, is there a way to use table extraction to grab the "products affected" correctly as well? In the above example, the 800mg gets appended to Zemaira instead of being part of Aralast.

**Reply by arthurthlee** [reply]
Also, the Products Affected section isn't extracted correctly for a similar doc like this (page 3 of the attached doc): 
![image](https://user-images.githubusercontent.com/2807789/157935649-2f65629e-a1e0-428a-aef5-29443e1a9311.png)
[doc5.pdf](https://github.com/jsvine/pdfplumber/files/8234784/doc5.pdf)

**Reply by jsvine** [reply]
Fun example! `doc5.pdf` is a great demonstration of how things can *look* similar to the naked while being encoded in the PDF quite differently. In this case, the line above "Products Affected" *looks* like the one in the other PDF, but is actually encoded as a line rather than a rectangle. No big problem, `pdfplumber` handles this easily. But more problematically, it's encoded as a *sliiiiiightly askew* line, rather than properly horizontal. You can see this by examining `p.lines`:

```
[{'x0': 72.0,
  'y0': 683.22,   # <- compare
  'x1': 540.0,
  'y1': 683.46,  # <- compare
  'width': 468.0,
  'height': 0.2400000000000091,
  'pts': [(72.0, 683.46), (540.0, 683.22)],
  'linewidth': 1,
  'stroke': True,
  'fill': False,
  'evenodd': False,
  'stroking_color': (0, 0, 0),
  'non_stroking_color': 0,
  'object_type': 'line',
  'page_number': 3,
  'top': 108.53999999999996,   # <- or alternatively, compare
  'bottom': 108.77999999999997,   # <- versus this
  'doctop': 1692.54}]
```

Because of this `pdfplumber` excludes it from the table-recognition algorithm, since tables are typically only composed of perfectly horizontal and vertical lines. (In fact, this is the first time I've seen an example of this, in many years of extracting tables from PDFs. Congrats on discovering new edge-cases!)

The simplest fix (rather than add heuristics for "is this *supposed* to be a horizontal line?") is probably to tweak the extraction settings to the following:

```python
p.extract_table({
    "explicit_vertical_lines": [ min(rect_x0s), max(rect_x1s) ],
    "explicit_horizontal_lines": [ line["top"] for line in p.lines ],
})
```

... which produces:

```
[['Products Affected\n'
  '\uf0b7 ABILIFY MAINTENA  \uf0b7 ABILIFY MAINTENA \n'
  'INTRAMUSCULAR  INTRAMUSCULAR \n'
  'SUSPENSION,EXTENDED REL  SUSPENSION,EXTENDED REL \n'
  'RECON 300 MG, 400 MG  SYRING',
  None],
 ['PA Criteria', 'Criteria Details'],
 ['Exclusion \nCriteria', ''],
 ['Required \nMedical \nInformation',
  'Documented history of receiving oral aripiprazole without any clinically \n'
  'significant side effects.  Failure or clinically significant adverse '
  'effects to \n'
  'two of the formulary alternatives: Invega Sustenna, Invega Trinza or \n'
  'Risperdal Consta.'],
 ['Age Restrictions', ''],
 ['Prescriber \nRestrictions', ''],
 ['Coverage \nDuration', 'Until the end of calendar year.'],
 ['Other Criteria', ''],
 ['Indications', 'All FDA-approved Indications.'],
 ['Off Label Uses', '']]
```

As for parsing the "Products Affected" sections themselves, that will probably take some custom code, since the two examples you've shared seem to format that section somewhat differently. You might have some luck passing the cell coordinates (found via `p.find_tables(...)[0].cells[0].bbox`) to `p.crop(...)`, and then using the experimental `.extract_text(layout=True)` on the cropped section. That would give you something a little more structured to try parsing.

**Reply by arthurthlee** [reply]
Thank you for the help! I've gone with using the text extraction strategy for now in order to parse the Products Affected section, but I'll try this suggestion out to see if it generalizes with the other PDFs that I'm working on.

One last question, the following PDF looks very similar to the above, but it seems like no rectangles are detected at all (tried by using the debug tablefinder in a notebook). Why would this be? I've fixed some other PDFs by changing the intersection tolerance, but I'm a bit stumped about this one: 

[doc112.pdf](https://github.com/jsvine/pdfplumber/files/8250162/doc112.pdf)

**Reply by jsvine** [reply]
Ah, yes. This PDF seems to represent its lines as `line` objects, rather than `rect`s. (Or, more correctly/explicitly: It draws its lines in formations that `pdfminer.six` [the library `pdfplumber` uses to extract this information] recognizes as `line` objects.) The good news is that you can grab *both* lines and the edges of rectangles through the `page.edges` property.  Should be relatively interchangeable with `.rects` in your code.

**Comment by yuchixing** [comment]
Hi, @jsvine , how to adjust the correct identification, thanks
![1665404714031](https://user-images.githubusercontent.com/2059565/194865918-0ed8f4a2-6e40-41fd-ae54-ebca1c81f065.png)
![1665405090533](https://user-images.githubusercontent.com/2059565/194866892-a40f3a2e-b3dc-4148-a908-4cc9a80ec0db.png)

**Reply by samkit-jain** [reply]
@yuchixing Please share the PDF as well after redacting any sensitive information so that we can help you better.

**Reply by yuchixing** [reply]
http://static.cninfo.com.cn/finalpage/2022-04-26/1213106404.PDF, page:146

**Reply by samkit-jain** [reply]
https://github.com/jsvine/pdfplumber/issues/488#issuecomment-894077855 does this help? The PDF looks similar to the one in the discussion.

**Reply by yuchixing** [reply]
Thans very much...3, ha ha
